app/components/SignUpForm.tsx

```diff
@@ -3,43 +3,71 @@ import { useState } from "react";
 import Input from "./Input";
 import Button from "./Button";
 import { User, Letter, Lock, LoadingSpinner, Confetti } from "./Icons";
+import { useForm } from "react-hook-form";
+import { FormData } from "../types";
 
 export default function SignUpForm() {
-  const [formData, setFormData] = useState({
-    email: "",
-    username: "",
-    password: "",
-    confirmPassword: "",
-  });
+  const {
+    register,
+    handleSubmit,
+    formState: { errors },
+    getValues,
+    watch,
+  } = useForm<FormData>();
   const [loading, setLoading] = useState(false);
   const [completed, setCompleted] = useState(false);
+  const hasErrors = Object.keys(errors).length > 0;
 
-  const handleChange = (e: React.ChangeEvent<HTMLInputElement>): void => {
-    const { name, value } = e.target;
-    setFormData((prev) => ({
-      ...prev,
-      [name]: value,
-    }));
-  };
-
-  const handleSubmit = (): void => {
-    setLoading(true);
-    setTimeout(() => {
-      setCompleted(true);
-      setLoading(false);
-    }, 1000);
+  const signUpRules = {
+    email: {
+      required: { value: true, message: "Email is required" },
+      pattern: {
+        value: /^[\w-.]+@([\w-]+\.)+[\w-]{2,4}$/,
+        message: "Invalid email address",
+      },
+    },
```

Comment:
Could you take a look into [zod](https://zod.dev/), more specifically its [rhf resolver](https://github.com/react-hook-form/resolvers?tab=readme-ov-file#zod)

Reply:
+1 and you can use then the `type FormData = z.infer<typeof formDataSchema>;`
https://zod.dev/?id=type-inference